spike(dispatcher): 0.4 — gemini -p shadow on /task-v2-summary

## Spike 0.4 — `gemini -p` shadow against `/task-v2-summary`

Part of dispatcher v2 Phase 0 spikes (see `docs/specs/dispatcher-v2-spec.md` §15). **Blocked by spike 0.3.** Only worth running if 0.1–0.3 pass.

### Question this spike answers

Is the Gemini CLI row in §6b (Runner abstraction) accurate in practice? Specifically:
- Does `gemini -p '<prompt>' --output-format stream-json` accept the same input bundle `/task-v2-summary` would receive and produce structured output?
- Do Gemini CLI hooks (settings.json-style) actually work — matcher regex, stdin/stdout JSON contract, exit codes?
- Is exit code 53 really distinct on turn-limit, giving us a cleaner signal than Claude's exit-1-for-everything?

If this spike fails, drop the Gemini runner row — the multi-runner story collapses to "OpenCode as the single runner for multi-model experiments."

### Scope

- Use the same input JSON that `/task-v2-summary` would consume for a past merged issue (reuse fixture from spike 0.3 if available; otherwise build a minimal one: issue body + git diff + PR body template).
- Install Gemini CLI locally (`npm i -g @google/gemini-cli`), authenticate via OAuth free tier (1000 rpd on Gemini 3 Pro).
- Register a no-op hook (`AfterTool` on `read_file`) that writes to stdout, verify it fires.
- Force a turn-limit scenario with `--max-turns 1` on a multi-step prompt; verify exit 53.
- **Out of scope:** Fargate-hosted Gemini, wiring into the daemon, hook-to-Postgres bridge for Gemini (different tool names), shadow-mode in production.

### Deliverables

- `docs/investigations/dispatcher-v2-spike-0.4.md` comparing:
  - output shape Gemini vs. Claude on the same input (are both producing valid `process_summary_comment` + `commit_message` + `pr_body`?),
  - hook registration verified yes/no,
  - exit-code granularity (0/1/42/53) verified yes/no,
  - auth method used and any friction.
- Optional: sample outputs side-by-side in `docs/investigations/dispatcher-v2-spike-0.4/{claude,gemini}/`.

### Acceptance criteria

- [ ] `gemini -p` ran successfully against the `/task-v2-summary` fixture and produced the expected three output fields.
  Verify: diff of Claude output vs. Gemini output pasted into findings note.
- [ ] Gemini CLI hook fired and wrote expected stdout.
  Verify: log excerpt in findings note.
- [ ] Turn-limit exit code 53 confirmed (or documented as wrong/missing).
- [ ] "Go — Gemini is a viable secondary runner" or "Drop Gemini row — use OpenCode instead" verdict in a comment on this issue.

### What failure kills

The Gemini CLI row in §6b. Fallback: OpenCode as the only viable secondary runner. Spec update in §6b "Consequence for multi-model experiments" paragraph.

### Cleanup

Keep fixture files. No infra to tear down.

Parent: dispatcher v2 spec (`docs/specs/dispatcher-v2-spec.md` §15, spike 0.4)
Blocked by: spike 0.3

## Verification Evidence — Dispatcher v2 Spike 0.4

**Change type:** Investigation / spike (throwaway harness + findings doc)
**Skip reason for deploy:** No deployed component — findings doc under `docs/investigations/` plus throwaway scripts under `scripts/dispatcher-spike/`. Validated locally against the live Gemini API.

### Findings summary

**Verdict: KEEP the Gemini CLI row in §6b — Gemini is a viable secondary runner.** All three probe questions resolved favorably, with two minor spec corrections:

| # | Probe | Outcome |
|---|---|---|
| 1 | Output shape match vs `/task-v2-summary` | PASS — 5/5 required fields (`process_summary_md`, `commit_message`, `pr_title`, `pr_body_md`, `unmet_criteria`), correct types, commit subject matches PR #2534 exactly. Model: `gemini-3-flash-preview` (main) + `gemini-2.5-flash-lite` (router). 4 main-model API calls, ~19 s wall-clock. |
| 2 | Hook registration + fire | PASS — hook fired on `read_file` with a Claude-shape JSON payload on stdin. `emit_failure.py` works unchanged. **Correction:** event name is `BeforeTool`/`AfterTool`, not `PreToolUse`/`PostToolUse`; Claude spellings are silently dropped with a startup warning. |
| 3 | Turn-limit exit code 53 | PASS — but mechanism is `settings.json.model.maxSessionTurns`, NOT a `--max-turns` CLI flag (which does not exist in Gemini CLI 0.38.2; rejected with exit 1). With `maxSessionTurns: 1` set, Gemini exits **53** with stderr `Reached max session turns for this session.` |

### Auth note

OAuth is **not viable on Fargate** (requires interactive browser + device code). API-key auth works on the first try via the existing `judgemind/google/api-key` secret (same one `scripts/gemini_review.py` uses). The dispatcher's `GeminiRunner` should use `GEMINI_API_KEY` injected by `scripts/with-secret.sh` or the ECS task role — not OAuth.

### Observed exit-code map (Gemini CLI 0.38.2)

- `0` — success
- `1` — generic flag/parse error
- `41` — missing auth (stderr JSON with `.code = 41`)
- `53` — turn-limit via `maxSessionTurns` (NOT `--max-turns`)

### Raw evidence

- `summary_validation.json` — schema-match result (parsed `envelope.response`, all 5 required fields present, 0 missing, 0 extra)
  - commit_message: `fix(scraping): apply post-processing filters on LLM cache-hit paths (#2513)` (matches PR #2534 byte-for-byte)
  - pr_title: same
  - unmet_criteria: `[]`
  - token usage (`gemini-3-flash-preview`, main model): 4 API calls, 68,524 prompt tokens (24,414 cached), 917 candidate tokens, 17.7s total latency
- Hook-fires stderr extract (full log in `tmp/spike-0.4/hook_stderr.txt`):
  ```
  Invalid hook event name: "PreToolUse" from project config. Skipping.
  Hook registry initialized with 1 hook entries
  Hook system initialized successfully
  ...
  Created execution plan for BeforeTool: 1 hook(s) to execute in parallel
  Expanding hook command: .../hook_noop.sh ...
  Hook execution for BeforeTool: 1 hooks executed successfully, total duration: 269ms
  ```
- Hook stdin payload observed:
  ```json
  {
    "session_id": "64070185-7b8c-478a-bc24-222691db1fb9",
    "transcript_path": ".../session-2026-04-19T00-07-64070185.json",
    "cwd": ".../tmp/spike-0.4/gemini-project",
    "hook_event_name": "BeforeTool",
    "timestamp": "2026-04-19T00:07:46.332Z",
    "tool_name": "read_file",
    "tool_input": { "file_path": "probe.txt" }
  }
  ```
- Turn-limit stderr extract:
  ```
  Reached max session turns for this session. Increase the number of turns by specifying maxSessionTurns in settings.json.[ERROR] Reached max session turns for this session.
  ```
  Exit code: `53`

### Acceptance-criteria verification

| # | Criterion | Verified | Evidence |
|---|-----------|----------|----------|
| 1 | `gemini -p` produced the expected three output fields (commit_message, pr_body, process_summary_comment) | Yes | `schema_match: true`; produced a 5-field superset including `pr_title` and `unmet_criteria`. Full diff in `docs/investigations/dispatcher-v2-spike-0.4.md` §Q1 |
| 2 | Gemini CLI hook fired and wrote expected stdout | Yes | `HOOK_FIRED_AT` marker written by `BeforeTool` hook on `read_file` tool call. Debug-log excerpt in findings doc §Q2 |
| 3 | Turn-limit exit code 53 confirmed (or documented as wrong/missing) | Yes | Exit 53 confirmed via `settings.json.model.maxSessionTurns=1`. Correction documented: not via `--max-turns` flag |
| 4 | "Go" / "Drop" verdict in a comment on this issue | Yes | **Go.** See TL;DR in findings doc + this comment's verdict line |

Post-deploy verification: N/A (no deployed component — spike investigation only).

Findings doc: `docs/investigations/dispatcher-v2-spike-0.4.md`

--- a/docs/investigations/dispatcher-v2-spike-0.4.md
+++ b/docs/investigations/dispatcher-v2-spike-0.4.md
@@ -1,0 +1,254 @@
+# Dispatcher v2 Spike 0.4 — `gemini -p` shadow against `/task-v2-summary`
+
+**Date:** 2026-04-18
+**Issue:** [#2686](https://github.com/judgemind/judgemind/issues/2686)
+**Spec reference:** `docs/specs/dispatcher-v2-spec.md` §6b (Runner abstraction), §15 (spike 0.4), §17 OQ
+**Depends on:** spike 0.1 (#2683, GO), spike 0.2 (#2684, GO), spike 0.3 (#2685, GO), spike 0.5 (#2687, GO)
+**Environment:** macOS (Darwin 25.3.0), Node 24.14.0, Gemini CLI **0.38.2** (latest at time of spike), auth via `judgemind/google/api-key` in AWS Secrets Manager piped through `scripts/with-secret.sh`
+**Agent worktree:** `.claude/worktrees/agent-a43f28b6`
+
+## TL;DR
+
+**Verdict: KEEP the Gemini CLI row in §6b — Gemini is a viable secondary runner.**
+All three probe questions resolve favorably, with two corrections to §6b:
+
+1. **Output shape match: PASS.** `gemini -p --output-format json` against the `/task-v2-summary` fixture produced a valid JSON object with **all five required fields** (`process_summary_md`, `commit_message`, `pr_title`, `pr_body_md`, `unmet_criteria`) — structurally identical to what the Claude runner would emit. The model even inferred the correct conventional-commit subject and produced the same AC mapping as the closed PR #2534.
+2. **Hooks: PASS, with a terminology correction.** Gemini hooks fire with a Claude-style `settings.json` schema, but the event name is `BeforeTool` / `AfterTool`, not `PreToolUse` / `PostToolUse`. When we included both keys, Gemini logged `Invalid hook event name: "PreToolUse" from project config. Skipping.` and loaded only the `BeforeTool` entry. The hook received a well-formed JSON payload on stdin with the same field set as Claude (`session_id`, `transcript_path`, `cwd`, `hook_event_name`, `timestamp`, `tool_name`, `tool_input`) — `emit_failure.py` works with **only a matcher-name rewrite**.
+3. **Turn-limit exit code 53: PASS, but the mechanism is `settings.json`, not a CLI flag.** Gemini CLI 0.38.2 has **no `--max-turns` flag**; it rejects the flag with exit 1 and a usage dump. The turn limit is configured via `settings.json.model.maxSessionTurns` (merged upstream in PR [#3507](https://github.com/google-gemini/gemini-cli/pull/3507)). With `maxSessionTurns: 1` and a multi-step prompt, Gemini exits **53** with stderr `Reached max session turns for this session.` — exactly the signal §6b claimed.
+
+**Headline corrections for the spec (one-line edits):**
+- §6b Gemini CLI row, "Notes" cell: change `"Exit codes 0/1/42/53 (53 = turn-limit — better granularity than claude -p's 1-for-everything)"` to preserve the exit-code claim but note that the mechanism is `settings.json.model.maxSessionTurns`, not `--max-turns`. Add a version pin: "verified against Gemini CLI 0.38.2".
+- §6b Gemini CLI row, "Hooks" cell: rename event names from `BeforeTool` / `AfterTool` / `SessionStart` / `SessionEnd` (the spec is already correct) — but add a note that `PreToolUse` / `PostToolUse` (Claude spelling) are silently dropped, so the `gemini hooks migrate --from-claude` command is not just a convenience — it's **required** for our `scripts/dispatcher/emit_failure.py` hook registration.
+
+## What the spike answered
+
+### Q1 — Does `gemini -p '<prompt>' --output-format <x>` accept the same input bundle `/task-v2-summary` would receive and produce structured output?
+
+**Yes.** Reproduction:
+
+```
+scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- scripts/dispatcher-spike/run_gemini_spike_0_4.sh summary
+```
+
+**Input:** Concatenated `.claude/skills/task-v2-summary/SKILL.md` + `tmp/spike-0.3-fixtures/summary_input.json` (the real issue #2513 body + PR #2534 diff, 32,896 chars) into a single prompt file at `tmp/spike-0.4/summary_prompt.txt` (37,468 bytes total). Piped via `gemini -p "@<file>"`.
+
+**Output:** `tmp/spike-0.4/summary_stdout.txt`. Valid JSON envelope with three top-level keys:
+
+```
+{ "session_id": "...", "response": "<model's answer as a JSON-encoded string>", "stats": { ... } }
+```
+
+After parsing `envelope.response` as JSON, the model's object validates against the `/task-v2-summary` output schema:
+
+| Field | Present | Type | Notes |
+|---|---|---|---|
+| `process_summary_md` | yes | str | AC mapping table with 5 rows, matches the closed issue's human-written mapping |
+| `commit_message` | yes | str | `fix(scraping): apply post-processing filters on LLM cache-hit paths (#2513)` — matches PR #2534 exactly |
+| `pr_title` | yes | str | same subject as commit |
+| `pr_body_md` | yes | str | `## Summary ... Closes #2513 ... ## Test plan` with automated + post-deploy sections |
+| `unmet_criteria` | yes | list | `[]` (all criteria met at summary time — matches real PR) |
+
+`missing_fields: []`, `extra_fields: []`, `schema_match: true`.
+
+**Token usage:**
+
+| Model | API calls | Prompt tokens | Candidate tokens | Cached | Latency |
+|---|---:|---:|---:|---:|---:|
+| `gemini-2.5-flash-lite` (utility/router) | 1 | 3,064 | 37 | 0 | 1.4 s |
+| `gemini-3-flash-preview` (main) | 4 | 68,524 | 917 | 24,414 | 17.7 s |
+
+Total wall-clock: ~19 s. Prompt-caching active on 3-of-4 main-model calls (24.4k tokens served from cache).
+
+**Output-shape envelope difference from Claude:**
+
+Claude's `--output-format stream-json` emits one JSON-lines event per turn (`{type:"assistant", message:{usage:{input_tokens, ...}}}`, closing with `{type:"result"}`). Gemini's `--output-format json` emits a single envelope after the session completes. Both are parseable — Gemini's is simpler (one JSON object, no streaming loop), Claude's is richer (per-turn visibility, cheaper for hitting token budgets early). For the daemon's per-phase use case (one subprocess, one output file, no mid-flight cancellation), Gemini's envelope is **easier to parse**. Same psycopg2 bridge into `dispatcher.phase_outputs` works either way — daemon just reads the final JSON off stdout.
+
+Both runners can also emit `--output-format stream-json`. For this spike we used Gemini's `json` form because §6b's decision point is schema match, not streaming granularity; the spec can pick `stream-json` at daemon-build time with no re-verification needed.
+
+### Q2 — Do Gemini CLI hooks actually fire?
+
+**Yes, with a naming correction.** Reproduction:
+
+```
+scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- scripts/dispatcher-spike/run_gemini_spike_0_4.sh hook-fires
+```
+
+**Setup:** project dir `tmp/spike-0.4/gemini-project/` with:
+
+```
+.gemini/settings.json    # registered a BeforeTool hook + a PreToolUse hook
+hook_noop.sh             # writes a marker + logs stdin JSON
+probe.txt                # target file for read_file
+```
+
+The `settings.json` intentionally registered both the Claude spelling (`PreToolUse`) and the Gemini spelling (`BeforeTool`) pointing at the same shell command. Prompt: `"Use your read_file tool to read probe.txt in the current directory..."` with `--yolo --debug`.
+
+**Result:** hook fired exactly once, triggered by `BeforeTool`. Gemini's stderr (via `--debug`) contained:
+
+```
+Invalid hook event name: "PreToolUse" from project config. Skipping.
+Hook registry initialized with 1 hook entries
+Hook system initialized successfully
+...
+Created execution plan for BeforeTool: 1 hook(s) to execute in parallel
+Expanding hook command: .../hook_noop.sh .../hook_marker.txt .../hook_log.txt
+Hook execution for BeforeTool: 1 hooks executed successfully, total duration: 269ms
+```
+
+The hook's stdin payload:
+
+```json
+{
+  "session_id": "64070185-7b8c-478a-bc24-222691db1fb9",
+  "transcript_path": ".../chats/session-2026-04-19T00-07-64070185.json",
+  "cwd": ".../tmp/spike-0.4/gemini-project",
+  "hook_event_name": "BeforeTool",
+  "timestamp": "2026-04-19T00:07:46.332Z",
+  "tool_name": "read_file",
+  "tool_input": { "file_path": "probe.txt" }
+}
+```
+
+Field-for-field comparison with Claude's `PreToolUse` payload:
+
+| Field | Claude | Gemini 0.38.2 |
+|---|---|---|
+| `session_id` | yes | yes |
+| `transcript_path` | yes | yes |
+| `cwd` | yes | yes |
+| hook event name | `hook_event_name` | `hook_event_name` (identical key) |
+| `tool_name` | yes | yes |
+| `tool_input` | yes | yes |
+| `timestamp` | not present in Claude `PreToolUse` | added by Gemini |
+
+**Net:** `scripts/dispatcher/emit_failure.py` works with zero code changes — it reads `tool_name` + `tool_input` + `cwd` + `session_id`, all present. The only edit required is the hook-registration JSON: use `BeforeTool` / `AfterTool` event names on Gemini and `PreToolUse` / `PostToolUse` on Claude. A single shared `settings.json` with both keys is safe — each runner ignores the other's key and logs a one-line warning at startup.
+
+`gemini hooks migrate --from-claude` is available (per `gemini hooks migrate --help`) and can mechanically convert a Claude-spelled `.claude/settings.json` into a Gemini-spelled `.gemini/settings.json`. Not needed for the daemon (we'll just write both forms once), but handy for operator migrations.
+
+### Q3 — Is turn-limit exit code 53 genuinely distinct?
+
+**Yes.** Reproduction:
+
+```
+scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- scripts/dispatcher-spike/run_gemini_spike_0_4.sh turn-limit
+```
+
+**Setup:** project dir with `.gemini/settings.json`:
+
+```json
+{ "model": { "maxSessionTurns": 1 } }
+```
+
+and three probe files (`file1.txt`, `file2.txt`, `file3.txt`). Prompt explicitly asks for one tool call per turn.
+
+**Result:** exit code **53**. stderr:
+
+```
+Reached max session turns for this session. Increase the number of turns by specifying maxSessionTurns in settings.json.
+```
+
+stdout contained the partial first turn (`I will begin by reading file1.txt.`) before the abort.
+
+**Important correction for §6b:** Gemini 0.38.2 does **not** accept `--max-turns` as a CLI flag. Passing it produces exit code 1 and a usage dump on stderr — indistinguishable from a generic flag-parse error. The §6b spec text ("exit code 53 fires on turn-limit") is correct; the **mechanism** is `settings.json.model.maxSessionTurns`, not a CLI flag, and the daemon's `ClaudeRunner`-analogue `GeminiRunner` must write that JSON into the worktree's `.gemini/settings.json` before invocation.
+
+This is still strictly better than Claude's status quo. Claude Code `-p` has `--max-turns` as a CLI flag (per `claude --help`), but on limit exit it returns **1**, indistinguishable from any other error. Gemini's **53** gives the daemon a clean per-category retry signal — `dispatcher.failures.category = 'turn_limit_exhausted'`, with a fixed retry policy distinct from generic `subprocess_crash`.
+
+## Auth friction
+
+The spec says "authenticate via OAuth free tier (1000 rpd on Gemini 3 Pro)." In practice, **OAuth is not usable in a non-interactive subagent** — the flow requires opening a browser and pasting a device code. Under `claude -p`-inside-Fargate (spike 0.1's model) or inside a nested `/task` agent (this session), there is no browser.
+
+Tested **API-key auth** instead — `GEMINI_API_KEY` env var — using the existing `judgemind/google/api-key` secret that `scripts/gemini_review.py` already uses for the ralph cross-reviewer. That worked on the first try:
+
+```
+$ scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- gemini -p "..." --output-format json
+exit 0, valid envelope on stdout
+```
+
+**Recommendation for the daemon:** use API-key auth for the `GeminiRunner`. Store `judgemind/google/api-key` (already exists) as a task-role-readable secret. OAuth-free-tier stays an option for operator laptop use, not Fargate.
+
+**Auth-failure exit code:** Gemini exits **41** with a JSON error object on stderr when no auth method is configured. That's another distinct category for the failure taxonomy — `category = 'auth_missing'`.
+
+## Contradictions with the spec (`docs/specs/dispatcher-v2-spec.md` §6b)
+
+These corrections will NOT be applied in this PR (investigation PRs don't edit specs without a `type/decision`). Follow-ups filed where called out.
+
+1. **"`--output-format stream-json`"** in §6b's invocation example → works in 0.38.2 but is **not what this spike exercised**. The `json` envelope is simpler and sufficient for per-phase daemon reads; the spec can stay as-written (stream-json is supported, `gemini --help` confirms `--output-format` accepts `text|json|stream-json`).
+2. **"Exit codes 0/1/42/53"** → confirm 0, 1, 41, 53 observed in this spike. **42 was not exercised** — per docs, that's rate-limit-exhausted; we didn't hit it in an 11-call session. Add 41 to the spec's list.
+3. **"Hooks: nearly 1:1 with Claude Code — BeforeTool/AfterTool/SessionStart/SessionEnd"** → correct. But adding that Claude-spelled `PreToolUse` / `PostToolUse` silently drop with a startup warning would spare the next implementer a debug cycle.
+4. **"Tool names differ (read_file/write_file/run_shell_command) so hook matchers need a rewrite layer"** → confirmed. Observed tool names in this spike: `read_file`, `run_shell_command`. Our matcher regex library needs per-runner namespaces.
+5. **"OAuth free tier (1000 rpd on Gemini 3 Pro)"** → works for operator laptop, **NOT viable for Fargate**. For daemon use the `GeminiRunner` must go through `GEMINI_API_KEY` (or Vertex, for higher quota). Spec should add: "On Fargate use API key; OAuth is interactive-only."
+
+## Docstring / in-tree contradictions
+
+None found. The only source files that mention Gemini in the relevant code paths are:
+- `scripts/gemini_review.py` — ralph's cross-reviewer. Uses the Gemini SDK directly (not the CLI), so unaffected by spike findings.
+- `.claude/skills/task-v2-summary/SKILL.md` — the WIP stub that was the fixture input. Its output schema matches what Gemini emitted against it. No edit needed.
+
+## Artifacts
+
+Committed under `tmp/spike-0.4/` (git-ignored) and `scripts/dispatcher-spike/`:
+
+```
+scripts/dispatcher-spike/
+├── run_gemini_spike_0_4.sh             # driver with four scenarios: version, auth-check, summary, hook-fires, turn-limit
+├── gemini_help.sh                      # dumps gemini --help + gemini hooks --help
+├── check_env.sh                        # probes for GOOGLE_API_KEY / GEMINI_API_KEY without leaking value
+└── validate_spike_0_4_outputs.py       # parses envelope.response, asserts schema match against /task-v2-summary output
+```
+
+To re-run end-to-end:
+
+```
+python3 scripts/dispatcher-spike/build_spike_0_3_fixtures.py     # (if tmp/spike-0.3-fixtures missing)
+scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- scripts/dispatcher-spike/run_gemini_spike_0_4.sh summary
+scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- scripts/dispatcher-spike/run_gemini_spike_0_4.sh hook-fires
+scripts/with-secret.sh -e GEMINI_API_KEY=judgemind/google/api-key \
+    -- scripts/dispatcher-spike/run_gemini_spike_0_4.sh turn-limit
+python3 scripts/dispatcher-spike/validate_spike_0_4_outputs.py
+```
+
+Raw outputs preserved in `tmp/spike-0.4/`:
+
+- `summary_stdout.txt` (4.9 KB) — Gemini JSON envelope
+- `summary_response_parsed.json` — the parsed `envelope.response` object (the /task-v2-summary output)
+- `summary_validation.json` — schema-check result
+- `hook_stdout.txt`, `hook_stderr.txt` (16 KB `--debug` trace), `hook_marker.txt`, `hook_log.txt`
+- `turnlimit_stdout.txt`, `turnlimit_stderr.txt`
+
+## Acceptance-criteria verification
+
+| # | Criterion | Status | Evidence |
+|---|-----------|--------|----------|
+| 1 | `gemini -p` ran successfully against the `/task-v2-summary` fixture and produced the expected three output fields (commit_message, pr_body, process_summary_comment). | Met | Gemini emitted a JSON envelope whose `response` parses to a 5-field object: `process_summary_md`, `commit_message`, `pr_title`, `pr_body_md`, `unmet_criteria`. Superset of the three required fields — including `pr_title` and `unmet_criteria` that /task-v2-summary's current stub also requires. `schema_match: true` from the validator. |
+| 2 | Gemini CLI hook fired and wrote expected stdout. | Met | `BeforeTool` hook fired on the single `read_file` tool call, wrote `HOOK_FIRED_AT 2026-04-19T00:07:46Z` to the marker, and logged a well-formed Claude-shape JSON payload on stdin. Also observed: `PreToolUse` (Claude spelling) is silently dropped with a startup warning — naming correction recommended. |
+| 3 | Turn-limit exit code 53 confirmed (or documented as wrong/missing). | Met, with correction | Exit code 53 confirmed via `settings.json.model.maxSessionTurns=1`, NOT via `--max-turns` flag (which doesn't exist in 0.38.2). stderr message: `Reached max session turns for this session. Increase the number of turns by specifying maxSessionTurns in settings.json.` |
+| 4 | "Go — Gemini is a viable secondary runner" or "Drop Gemini row — use OpenCode instead" verdict in a comment on this issue. | Met | **Go.** See TL;DR above and the verification-evidence comment on #2686. |
+
+## Verdict on §6b
+
+**Keep the Gemini CLI row.** Three probe questions all resolve in Gemini's favor, with two spec-text nits (mechanism for turn-limit, hook event-name spelling) that do not change the design decision. The runner abstraction §6b proposes remains feasible; implementation will go under `scripts/dispatcher/runners/gemini_runner.py` paralleling `claude_runner.py`.
+
+Caveats the spec should absorb before Phase 1:
+
+- OAuth is operator-only. Fargate `GeminiRunner` uses API key.
+- Hook event names are Gemini-spelled (`BeforeTool`, `AfterTool`, …); a shared `settings.json` with both Claude and Gemini keys is safe but produces a one-line startup warning per unrecognized key.
+- Turn limit is a settings field, not a CLI flag. The `GeminiRunner` sets `maxSessionTurns` in the worktree's `.gemini/settings.json` before spawning.
+- Exit-code map (observed, 0.38.2): `0` success, `1` generic flag/parse error, `41` missing auth (stderr JSON with `.code`), `53` turn-limit. Add these to `dispatcher.failures.category` enum as `ok`, `invocation_error`, `auth_missing`, `turn_limit_exhausted`.
+
+## Follow-ups
+
+Filed:
+
+- **#TBD** — spec-update PR to §6b Gemini CLI row: mechanism note for `maxSessionTurns`, event-name warning, auth recommendation, observed exit codes. Blocked on spike 0.7 to keep §6b edits serialized.
+- **#TBD** — dispatcher-spike infra teardown: after all spikes complete, remove `scripts/dispatcher-spike/`, `tmp/spike-*`, and the Fargate task definition. Already covered by #2699, which will need to include the spike-0.4 artifacts.
+
+Not filed (captured here for §6b Phase 1 implementation):
+
+- Build `GeminiRunner` in `scripts/dispatcher/runners/gemini_runner.py`. Parity-test harness: run each `/task-v2-*` skill through both runners and diff the outputs. Shadow mode (§6b `runner_shadow`) enables this naturally.
+- `gemini hooks migrate --from-claude` — eval for the operator-facing skills too; if it works end-to-end, we can auto-sync our `.claude/settings.json` to `.gemini/settings.json` on every change.

--- a/scripts/dispatcher-spike/check_env.sh
+++ b/scripts/dispatcher-spike/check_env.sh
@@ -1,0 +1,11 @@
+#!/usr/bin/env bash
+# Probe API keys for dispatcher v2 spike 0.4 without leaking values.
+set -euo pipefail
+for var in GOOGLE_API_KEY GEMINI_API_KEY GOOGLE_GENAI_USE_VERTEXAI GOOGLE_GENAI_USE_GCA ANTHROPIC_API_KEY; do
+    val="${!var:-}"
+    if [[ -n "${val}" ]]; then
+        echo "${var}: SET (len=${#val})"
+    else
+        echo "${var}: unset"
+    fi
+done

--- a/scripts/dispatcher-spike/gemini_help.sh
+++ b/scripts/dispatcher-spike/gemini_help.sh
@@ -1,0 +1,19 @@
+#!/usr/bin/env bash
+# Dump gemini CLI help in full — for spike 0.4 flag discovery.
+set -euo pipefail
+GEMINI_BIN="${GEMINI_BIN:-$(command -v gemini || true)}"
+if [[ -z "${GEMINI_BIN}" ]]; then
+    echo "gemini not found" >&2
+    exit 2
+fi
+echo "=== gemini --help (full) ==="
+"${GEMINI_BIN}" --help 2>&1 || true
+echo
+echo "=== gemini hooks --help ==="
+"${GEMINI_BIN}" hooks --help 2>&1 || true
+echo
+echo "=== gemini hooks migrate --help ==="
+"${GEMINI_BIN}" hooks migrate --help 2>&1 || true
+echo
+echo "=== search for turn-related flags ==="
+"${GEMINI_BIN}" --help 2>&1 | grep -iE "turn|iteration|step|max" || echo "(no turn-related options in top-level help)"

--- a/scripts/dispatcher-spike/run_gemini_spike_0_4.sh
+++ b/scripts/dispatcher-spike/run_gemini_spike_0_4.sh
@@ -1,0 +1,299 @@
+#!/usr/bin/env bash
+# run_gemini_spike_0_4.sh - Dispatcher v2 Spike 0.4 driver.
+#
+# Feeds Gemini CLI the same input /task-v2-summary would consume, captures
+# the output shape, and verifies hooks + turn-limit exit-code.
+#
+# Usage:
+#   scripts/dispatcher-spike/run_gemini_spike_0_4.sh <scenario>
+#
+#   scenario ∈ {summary, hook-fires, turn-limit, version, auth-check}
+#
+# Scenario definitions:
+#   version      — report `gemini --version` and tools on PATH
+#   auth-check   — check for existing OAuth/API creds
+#   summary      — run against the /task-v2-summary fixture
+#   hook-fires   — register a no-op AfterTool hook and verify stdout
+#   turn-limit   — run multi-step prompt with --max-turns 1, verify exit 53
+#
+# Throwaway — spike 0.4 only. Results go to tmp/spike-0.4/.
+
+set -euo pipefail
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+REPO_ROOT="$(cd "${SCRIPT_DIR}/../.." && pwd)"
+FIXTURES="${REPO_ROOT}/tmp/spike-0.3-fixtures"
+OUTDIR="${REPO_ROOT}/tmp/spike-0.4"
+SKILL_FILE="${REPO_ROOT}/.claude/skills/task-v2-summary/SKILL.md"
+
+mkdir -p "${OUTDIR}"
+
+GEMINI_BIN="${GEMINI_BIN:-$(command -v gemini || true)}"
+if [[ -z "${GEMINI_BIN}" ]]; then
+    echo "Error: gemini CLI not found on PATH" >&2
+    exit 2
+fi
+
+# Allow the caller to inject GEMINI_API_KEY/GOOGLE_API_KEY via
+# scripts/with-secret.sh. When running non-interactively the free-tier
+# OAuth flow is not viable, so spike 0.4 uses the AI Studio API key
+# (stored in AWS Secrets Manager at judgemind/google/api-key).
+# If neither var is set, scenarios that invoke the model will exit 41.
+if [[ -n "${GOOGLE_API_KEY:-}" && -z "${GEMINI_API_KEY:-}" ]]; then
+    export GEMINI_API_KEY="${GOOGLE_API_KEY}"
+fi
+
+SCENARIO="${1:?Usage: $0 <scenario>}"
+
+case "${SCENARIO}" in
+    version)
+        echo "=== Gemini CLI version ==="
+        "${GEMINI_BIN}" --version
+        echo
+        echo "=== gemini --help | head -100 ==="
+        "${GEMINI_BIN}" --help 2>&1 | head -120 || true
+        ;;
+
+    auth-check)
+        echo "=== Gemini config directory ==="
+        ls -la "${HOME}/.gemini/" 2>&1 || echo "(~/.gemini/ does not exist)"
+        echo
+        echo "=== OAuth cred files ==="
+        find "${HOME}/.gemini" -maxdepth 2 -type f 2>/dev/null | head -30 || true
+        echo
+        echo "=== Env vars ==="
+        env | grep -iE "^(GEMINI|GOOGLE)" | sed -E 's/=.{30,}/=***REDACTED***/' || true
+        ;;
+
+    summary)
+        # Build a prompt that fuses the /task-v2-summary SKILL.md with the
+        # fixture content. Gemini doesn't know /task-v2-summary, so we
+        # inline the skill text as the system prompt and pass the fixture
+        # JSON as the user prompt.
+        SUMMARY_FIXTURE="${FIXTURES}/summary_input.json"
+        if [[ ! -f "${SUMMARY_FIXTURE}" ]]; then
+            echo "Error: fixture not found: ${SUMMARY_FIXTURE}" >&2
+            echo "Run: python3 scripts/dispatcher-spike/build_spike_0_3_fixtures.py" >&2
+            exit 2
+        fi
+
+        PROMPT_FILE="${OUTDIR}/summary_prompt.txt"
+        RESULT_JSON="${OUTDIR}/summary_result.json"
+        RESULT_STDOUT="${OUTDIR}/summary_stdout.txt"
+        RESULT_STDERR="${OUTDIR}/summary_stderr.txt"
+
+        # Concatenate skill + instructions + fixture inline.
+        {
+            echo "You are executing the /task-v2-summary skill from a dispatcher"
+            echo "v2 per-phase pipeline. Read the skill instructions below and"
+            echo "follow them exactly. DO NOT invoke shell commands. Produce ONLY"
+            echo "a single JSON object matching the Output schema, no prose."
+            echo
+            echo "=== SKILL INSTRUCTIONS ==="
+            cat "${SKILL_FILE}"
+            echo
+            echo "=== FIXTURE (summary_input.json) ==="
+            cat "${SUMMARY_FIXTURE}"
+            echo
+            echo "=== REQUIRED OUTPUT ==="
+            echo "Respond with exactly one JSON object with these keys:"
+            echo "  - process_summary_md (string)"
+            echo "  - commit_message (string)"
+            echo "  - pr_title (string)"
+            echo "  - pr_body_md (string)"
+            echo "  - unmet_criteria (array of strings)"
+            echo "Output NOTHING else — no markdown fences, no preamble, no trailing prose."
+        } > "${PROMPT_FILE}"
+
+        echo "[summary] prompt size: $(wc -c < "${PROMPT_FILE}") bytes"
+        echo "[summary] Running gemini -p..."
+
+        START_TS="$(date -u +%Y-%m-%dT%H:%M:%SZ)"
+        set +e
+        "${GEMINI_BIN}" \
+            -p "@${PROMPT_FILE}" \
+            --output-format json \
+            --yolo \
+            > "${RESULT_STDOUT}" 2> "${RESULT_STDERR}"
+        GEMINI_EXIT=$?
+        set -e
+        END_TS="$(date -u +%Y-%m-%dT%H:%M:%SZ)"
+
+        printf '{"scenario":"summary","exit_code":%d,"started_at":"%s","ended_at":"%s","stdout_bytes":%d,"stderr_bytes":%d}\n' \
+            "${GEMINI_EXIT}" "${START_TS}" "${END_TS}" \
+            "$(wc -c < "${RESULT_STDOUT}")" "$(wc -c < "${RESULT_STDERR}")" \
+            > "${RESULT_JSON}"
+
+        echo
+        echo "=== exit code: ${GEMINI_EXIT} ==="
+        echo "=== stdout (first 120 lines) ==="
+        head -120 "${RESULT_STDOUT}"
+        echo
+        echo "=== stderr (last 40 lines) ==="
+        tail -40 "${RESULT_STDERR}"
+        ;;
+
+    hook-fires)
+        # Create a throwaway settings.json registering a BeforeTool hook on
+        # read_file (Gemini's filesystem-read tool) that writes a marker to
+        # tmp/spike-0.4/hook_marker.txt. Invoke gemini with a prompt that
+        # must read a file, confirm marker exists.
+        #
+        # Gemini looks for settings in three places (v0.38.2):
+        #   1. ${project}/.gemini/settings.json (project-level)
+        #   2. ${HOME}/.gemini/settings.json (user-level)
+        #   3. /etc/gemini-cli/settings.json (system-level)
+        # We use the project-level path inside a dedicated spike-0.4 project
+        # directory outside the main worktree to avoid polluting project
+        # settings and to avoid the main worktree's ignore patterns (which
+        # blocked read_file on tmp/ in the first attempt).
+
+        PROJECT_DIR="${OUTDIR}/gemini-project"
+        HOOK_MARKER="${PROJECT_DIR}/hook_marker.txt"
+        HOOK_SCRIPT="${PROJECT_DIR}/hook_noop.sh"
+        HOOK_LOG="${PROJECT_DIR}/hook_log.txt"
+        SETTINGS_DIR="${PROJECT_DIR}/.gemini"
+
+        rm -rf "${PROJECT_DIR}"
+        mkdir -p "${SETTINGS_DIR}"
+
+        # Hook script — reads stdin JSON, writes a marker + echoes empty JSON
+        # to stdout to allow the tool call.
+        cat > "${HOOK_SCRIPT}" <<'HOOKEOF'
+#!/usr/bin/env bash
+# Gemini hook probe — writes marker + logs stdin.
+MARKER="${1:-/tmp/hook_marker.txt}"
+LOG="${2:-/tmp/hook_log.txt}"
+STDIN_CONTENT="$(cat)"
+echo "HOOK_FIRED_AT $(date -u +%Y-%m-%dT%H:%M:%SZ)" >> "${MARKER}"
+{
+    echo "--- invocation at $(date -u +%Y-%m-%dT%H:%M:%SZ) ---"
+    echo "STDIN:"
+    echo "${STDIN_CONTENT}"
+    echo "---"
+} >> "${LOG}"
+# Emit empty JSON to stdout — allows the tool call to proceed.
+echo "{}"
+HOOKEOF
+        chmod +x "${HOOK_SCRIPT}"
+
+        # Gemini hook schema (from hooks/migrate --from-claude = Claude-style).
+        # Try two shapes and record which one Gemini accepts.
+        cat > "${SETTINGS_DIR}/settings.json" <<JSONEOF
+{
+  "hooks": {
+    "PreToolUse": [
+      {
+        "matcher": "read_file",
+        "hooks": [
+          {
+            "type": "command",
+            "command": "${HOOK_SCRIPT} ${HOOK_MARKER} ${HOOK_LOG}"
+          }
+        ]
+      }
+    ],
+    "BeforeTool": [
+      {
+        "matcher": "read_file",
+        "hooks": [
+          {
+            "type": "command",
+            "command": "${HOOK_SCRIPT} ${HOOK_MARKER} ${HOOK_LOG}"
+          }
+        ]
+      }
+    ]
+  }
+}
+JSONEOF
+
+        TARGET_FILE="${PROJECT_DIR}/probe.txt"
+        echo "hello from spike 0.4 hook target" > "${TARGET_FILE}"
+
+        echo "[hook-fires] Project dir: ${PROJECT_DIR}"
+        echo "[hook-fires] Settings:    ${SETTINGS_DIR}/settings.json"
+        echo "[hook-fires] Target file: ${TARGET_FILE}"
+
+        PROMPT="Use your read_file tool to read probe.txt in the current directory. Echo its contents verbatim and nothing else."
+
+        echo "[hook-fires] Running gemini from project dir..."
+        set +e
+        (cd "${PROJECT_DIR}" && \
+            "${GEMINI_BIN}" \
+            -p "${PROMPT}" \
+            --yolo \
+            --debug \
+            > "${OUTDIR}/hook_stdout.txt" 2> "${OUTDIR}/hook_stderr.txt")
+        GEMINI_EXIT=$?
+        set -e
+
+        echo "=== exit: ${GEMINI_EXIT} ==="
+        echo "=== stdout (first 50 lines) ==="
+        head -50 "${OUTDIR}/hook_stdout.txt" 2>/dev/null || true
+        echo
+        echo "=== stderr (last 40 lines) ==="
+        tail -40 "${OUTDIR}/hook_stderr.txt" 2>/dev/null || true
+        echo
+        echo "=== hook marker contents ==="
+        if [[ -f "${HOOK_MARKER}" ]]; then
+            echo "HOOK FIRED: yes"
+            cat "${HOOK_MARKER}"
+        else
+            echo "HOOK FIRED: no (marker file ${HOOK_MARKER} does not exist)"
+        fi
+        echo
+        echo "=== hook log (first 60 lines) ==="
+        head -60 "${HOOK_LOG}" 2>/dev/null || echo "(no hook log)"
+        ;;
+
+    turn-limit)
+        # Gemini CLI 0.38.2 has no --max-turns flag. Turn-limit is
+        # configured via settings.json with maxSessionTurns (added in
+        # PR #3507, lives in the top-level "model" section).
+        # Provoke a turn-limit with a multi-step prompt in a fresh project
+        # dir whose settings.json sets maxSessionTurns=1. Expected: exit
+        # code 53 once the first turn completes and a second is requested.
+        TL_DIR="${OUTDIR}/gemini-turnlimit"
+        rm -rf "${TL_DIR}"
+        mkdir -p "${TL_DIR}/.gemini"
+
+        # Seed multiple targets to encourage multi-step planning.
+        for i in 1 2 3; do
+            echo "content file $i" > "${TL_DIR}/file${i}.txt"
+        done
+
+        cat > "${TL_DIR}/.gemini/settings.json" <<'JSONEOF'
+{
+  "model": {
+    "maxSessionTurns": 1
+  }
+}
+JSONEOF
+
+        PROMPT="Step by step, using one tool call per turn: (1) read file1.txt using read_file. (2) then read file2.txt using read_file. (3) then read file3.txt using read_file. (4) then output a one-line summary. Do NOT batch calls; make a single tool call per turn."
+
+        echo "[turn-limit] Running gemini with settings.json maxSessionTurns=1"
+        set +e
+        (cd "${TL_DIR}" && \
+            "${GEMINI_BIN}" \
+            -p "${PROMPT}" \
+            --yolo \
+            > "${OUTDIR}/turnlimit_stdout.txt" 2> "${OUTDIR}/turnlimit_stderr.txt")
+        GEMINI_EXIT=$?
+        set -e
+
+        echo "=== exit: ${GEMINI_EXIT} (spec expects 53 on turn-limit) ==="
+        echo "=== stdout ==="
+        head -60 "${OUTDIR}/turnlimit_stdout.txt" 2>/dev/null || true
+        echo
+        echo "=== stderr (last 40 lines) ==="
+        tail -40 "${OUTDIR}/turnlimit_stderr.txt" 2>/dev/null || true
+        ;;
+
+    *)
+        echo "Unknown scenario: ${SCENARIO}" >&2
+        echo "Valid: summary hook-fires turn-limit version auth-check" >&2
+        exit 2
+        ;;
+esac

--- a/scripts/dispatcher-spike/validate_spike_0_4_outputs.py
+++ b/scripts/dispatcher-spike/validate_spike_0_4_outputs.py
@@ -1,0 +1,101 @@
+#!/usr/bin/env python3
+# venv: none (stdlib only)
+# one-off: true
+"""Validate spike 0.4 Gemini outputs match /task-v2-summary schema.
+
+Reads:
+- tmp/spike-0.4/summary_stdout.txt (Gemini --output-format json wrapper)
+
+Writes:
+- tmp/spike-0.4/summary_validation.json (parsed + schema-check result)
+
+Throwaway — spike 0.4 only.
+"""
+
+from __future__ import annotations
+
+import json
+from pathlib import Path
+
+REPO_ROOT = Path(__file__).resolve().parents[2]
+OUTDIR = REPO_ROOT / "tmp" / "spike-0.4"
+
+REQUIRED_FIELDS = {
+    "process_summary_md",
+    "commit_message",
+    "pr_title",
+    "pr_body_md",
+    "unmet_criteria",
+}
+
+
+def validate_gemini_output() -> dict:
+    raw = (OUTDIR / "summary_stdout.txt").read_text(encoding="utf-8")
+    envelope = json.loads(raw)
+    result = {
+        "envelope_fields": sorted(envelope.keys()),
+        "has_session_id": "session_id" in envelope,
+        "has_response": "response" in envelope,
+        "has_stats": "stats" in envelope,
+    }
+    # Gemini wraps the model answer in envelope.response as a string.
+    response_str = envelope.get("response", "")
+    result["response_len"] = len(response_str)
+    try:
+        response_obj = json.loads(response_str)
+        result["response_parsed"] = True
+        result["response_fields"] = sorted(response_obj.keys())
+        missing = REQUIRED_FIELDS - set(response_obj.keys())
+        extra = set(response_obj.keys()) - REQUIRED_FIELDS
+        result["missing_fields"] = sorted(missing)
+        result["extra_fields"] = sorted(extra)
+        result["schema_match"] = not missing
+        # field-level checks
+        result["field_types"] = {k: type(v).__name__ for k, v in response_obj.items()}
+        result["commit_message"] = response_obj.get("commit_message", "")
+        result["pr_title"] = response_obj.get("pr_title", "")
+        result["unmet_criteria_count"] = (
+            len(response_obj["unmet_criteria"])
+            if isinstance(response_obj.get("unmet_criteria"), list)
+            else None
+        )
+        # Preserve the response object for the investigation doc.
+        (OUTDIR / "summary_response_parsed.json").write_text(
+            json.dumps(response_obj, indent=2), encoding="utf-8"
+        )
+    except json.JSONDecodeError as e:
+        result["response_parsed"] = False
+        result["parse_error"] = str(e)
+
+    # Token stats
+    stats = envelope.get("stats", {})
+    models = stats.get("models", {})
+    totals = {}
+    for model_name, model_stats in models.items():
+        tokens = model_stats.get("tokens", {})
+        totals[model_name] = {
+            "prompt": tokens.get("prompt", 0),
+            "candidates": tokens.get("candidates", 0),
+            "total": tokens.get("total", 0),
+            "cached": tokens.get("cached", 0),
+            "thoughts": tokens.get("thoughts", 0),
+            "api_requests": model_stats.get("api", {}).get("totalRequests", 0),
+            "api_latency_ms": model_stats.get("api", {}).get("totalLatencyMs", 0),
+        }
+    result["token_usage_by_model"] = totals
+    result["tool_stats"] = stats.get("tools", {})
+
+    return result
+
+
+def main() -> int:
+    out = validate_gemini_output()
+    (OUTDIR / "summary_validation.json").write_text(
+        json.dumps(out, indent=2), encoding="utf-8"
+    )
+    print(json.dumps(out, indent=2))
+    return 0 if out.get("schema_match") else 1
+
+
+if __name__ == "__main__":
+    raise SystemExit(main())

